### DISCUSSION TITLE
Inverted ABCI interface

### DISCUSSION THREAD

**Discussion by tac0turtle** [tendermint/spec#279] [Ideas]
Inverted ABCI is a secondary interface made for applications to submit transactions and query Tendermint state through. 

The current design of the ABCI treats Tendermint as a client and the application as the server. This allows Tendermint to send data to the application and the app to respond with the requested data. The way an application would get data into Tendermint is through the RPC. This allows apps to submit transactions to Tendermint and query Tendermint state.

### Current Transaction Submission Flow

When an application submits a transaction to Tendermint it goes through the below flow. 

- A user submits a transaction to the application.
- Application submits the transaction to Tendermint via the RPC.
- Tendermint calls checkTx, asking the application to verify the validity of transaction.
- Application confirms the tx is valid and responds with the Tx.
- Tendermint  adds the tx to the mempool. 

A user can submit the tx directly to Tendermint RPC as well, this will skip the first bullet point of the above list. 

I have run into many questions about why it is this way. I myself, when starting to work in the cosmos ecosystem, failed to understand this and many people still lack the understanding of this. While the understanding is not the problem, I would say the problem is that we are making unnecessary trips to the application in the above flow. We are also aware that the RPC is a DOS vector, but still ask applications to use it for submission of transactions.

### Proposed Transaction Submission Flow

If we were to expose a new protobuf service, which is optional, for applications to submit transactions directly to the Tendermint mempool we allow apps to avoid calling the RPC unnecessarily. This will avoid calling `CheckTx` on transaction submission.

#### New Service 

```proto
message RequestSubmitTx {
  uint32         code       = 1;
  bytes          data       = 2;
  string         log        = 3;  // nondeterministic
  string         info       = 4;  // nondeterministic
  int64          gas_wanted = 5 [json_name = "gas_wanted"];
  int64          gas_used   = 6 [json_name = "gas_used"];
  repeated Event events     = 7
      [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
}

message ResponseSubmitTx {
 error error = 1;
}

// ABCITendermint teats Tendermint as the server and the Application as the client 
service ABCITendermint {
 rpc SubmitTx(SubmitTxReq) returns (SubmitTxRes);
}
```

### Current Query Flow

> This section does not cover `abci_query` endpoint.

The current query flow is similar to the transaction submission flow with a few differences. The query for Tendermint data also goes through the RPC layer, except the application may wrap the returned data. 

### Proposed Query Flow

The new flow would go through the above-mentioned service. This part has some overlap with https://github.com/tendermint/spec/issues/203, but this section will be left empty as it requires more thought. 

Overlap with https://github.com/tendermint/tendermint/discussions/5977

short presentation used to explain this to the tendermint team: https://docs.google.com/presentation/d/1kzGhIm6YtN5OFocekwv2kXHGRo0s_7yMfx22PVtJJv4/edit?usp=sharing

**Comment by robert-zaremba** [comment]
`App` is not exposed publicly for few reasons:
* app doesn't have the network protection
* app is assumed to run in a privileged zone.

So, unless we will do more complex setup around the App, the extra round trip won't matter match because we in production 99.9999% of users will use Tendermint RPC rather than app.

**Reply by tac0turtle** [reply]
Not sure what you mean by `app is not exposed publicly for few reasons:`? Node operators many times expose gRPC, REST or grpc-gateway in the case of the sdk to the public internet, this is where transactions are submitting in some cases. 

The above proposal would tell node operators they can expose only a single endpoint instead of two or more.

**Reply by robert-zaremba** [reply]
Oh, I thought that operators won't expose gRPC openly - only to permission-ed users.

**Reply by tac0turtle** [reply]
Same could be said about RPC, but I have learned people are good at using software poorly. 😄

**Reply by robert-zaremba** [reply]
So they don't have any filtering to give an access to only subscribed users?
